pkg/core/db/db.go

```diff
@@ -2,24 +2,33 @@ package db
 
 import (
 	"context"
+	"database/sql"
+	_ "github.com/lib/pq"
 
 	"github.com/skygeario/skygear-server/pkg/core/config"
 )
 
-type DBProvider struct{}
-
-func (p DBProvider) Provide(ctx context.Context, tConfig config.TenantConfiguration) IDB {
-	return &DB{tConfig.DBConnectionStr}
-}
-
-type IDB interface {
-	GetRecord(string) string
+// DBProvider is providing a postgres DB connection instance from
+// connection poll that is ready to use by the tenant unware gear business
+// code.
+type DBProvider struct {
+	namespace string
 }
 
-type DB struct {
-	ConnectionStr string
+func NewDBProvider(gear string) *DBProvider {
+	return &DBProvider{
+		namespace: gear,
+	}
 }
 
-func (db DB) GetRecord(recordID string) string {
-	return db.ConnectionStr + ":" + recordID
+func (p DBProvider) Provide(tConfig config.TenantConfiguration) *sql.Conn {
+	db, err := sql.Open("postgres", tConfig.DBConnectionStr)
+	if err != nil {
+		panic(err)
+	}
+	conn, err := db.Conn(context.Background())
```

Comment:
I suggest passing the context as arg of this function. Thus when a handler need to provide a db connection, it can use the request context, and the db connection would be bound to the request life cycle, i.e. when a http request is finished, the db connection would be cleaned up by the underlying db library.

Reply:
agree